middleware/grpc_instrumentation_test.go

```diff
@@ -43,3 +55,260 @@ func TestErrorCode_Unknown(t *testing.T) {
+	err = stream.CloseSend()
+	require.Equal(t, mockStream.closeSendError, err)
+	requireRequestMetrics(t, reg, "error", 1)
+
+	requireDoesNotIncrementMetricsAfterCompletion(t, stream, mockStream, cancelCtx, reg, "error")
+}
+
+func TestStreamClientInstrumentInterceptor_ContextCancelled(t *testing.T) {
+	stream, mockStream, cancelCtx, reg := setUpStreamClientInstrumentInterceptorTest(t, true)
+
+	cancelCtx()
+
+	require.Eventually(t, func() bool {
+		return gatherRequestMetrics(t, reg) != ""
+	}, time.Second, 10*time.Millisecond, "gave up waiting for the context cancellation to report metrics")
+
+	requireRequestMetrics(t, reg, "cancel", 1)
+	requireDoesNotIncrementMetricsAfterCompletion(t, stream, mockStream, cancelCtx, reg, "cancel")
+}
+
+func requireDoesNotIncrementMetricsAfterCompletion(t *testing.T, stream grpc.ClientStream, mockStream *mockGrpcStream, cancelCtx context.CancelFunc, reg *prometheus.Registry, result string) {
+	// Should not increment metric, even if Header now returns an error.
+	mockStream.headerError = errors.New("calling Header() failed")
+	_, err := stream.Header()
+	require.Equal(t, mockStream.headerError, err)
+	requireRequestMetrics(t, reg, result, 1)
+
+	// Should not increment metric, even if SendMsg now returns an error
+	mockStream.sendMsgError = errors.New("calling SendMsg() failed")
+	err = stream.SendMsg("one last message")
+	require.Equal(t, mockStream.sendMsgError, err)
+	requireRequestMetrics(t, reg, result, 1)
+
+	// Should not increment metric, even if RecvMsg now returns an error
+	mockStream.recvMsgError = errors.New("calling RecvMsg() failed")
+	err = stream.RecvMsg(nil)
+	require.Equal(t, mockStream.recvMsgError, err)
+	requireRequestMetrics(t, reg, result, 1)
+
+	// Should not increment metric, even if RecvMsg now returns EOF
+	mockStream.recvMsgError = io.EOF
+	err = stream.RecvMsg(nil)
+	require.Equal(t, mockStream.recvMsgError, err)
+	requireRequestMetrics(t, reg, result, 1)
+
+	// Should not increment metric, even if CloseSend now returns an error
+	mockStream.closeSendError = errors.New("calling CloseSend() failed")
+	err = stream.CloseSend()
+	require.Equal(t, mockStream.closeSendError, err)
+	requireRequestMetrics(t, reg, result, 1)
+
+	// Should not increment metric, even if the context is cancelled
+	cancelCtx()
+	requireRequestMetrics(t, reg, result, 1)
+}
+
+func requireNoRequestMetrics(t *testing.T, g prometheus.Gatherer) {
+	actual := gatherRequestMetrics(t, g)
+	require.Empty(t, actual, "expected no request metrics to be reported")
+}
+
+func requireRequestMetrics(t *testing.T, g prometheus.Gatherer, result string, count int) {
+	expected := fmt.Sprintf(`client_request_duration_seconds_count{method="/thing.Server/DoThing", result="%v"} %v`, result, count)
+	actual := gatherRequestMetrics(t, g)
+
+	require.Equalf(t, expected, actual, "expected count of %v for requests with result '%v'", count, result)
+}
+
+func gatherRequestMetrics(t *testing.T, g prometheus.Gatherer) string {
```

Comment:
[non blocking] I was wondering if we could reuse the utilities from the `metrics` package, but I don't think we can out of the box. Not high priority, but I think in Mimir we have most of the missing utilities ready and we could move them to dskit:
- `findHistogramMatchingLabels()` from `pkg/frontend/querymiddleware/codec_json_test.go`
- `getMetricsMatchingLabels()` from `pkg/storegateway/bucket_e2e_test.go`
- `MatchesSelectors()` from `pkg/util`

Reply:
I agree, I've created https://github.com/grafana/dskit/issues/351 to track this